akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala

```diff
@@ -126,6 +126,36 @@ private[akka] final case class MonitorInterceptor[T](actorRef: ActorRef[T]) exte
 
 }
 
+/**
+ * Log all messages for this decorated ReceiveTarget[T] to logger before receiving it ourselves.
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final case class LogMessagesInterceptor[T](opts: LogOptions) extends BehaviorInterceptor[T, T] {
+  private val logger: Option[Logger] = Option(opts.logger)
+
+  import BehaviorInterceptor._
+
+  override def aroundReceive(ctx: TypedActorContext[T], msg: T, target: ReceiveTarget[T]): Behavior[T] = {
+    if (opts.enabled)
+      logger.getOrElse(ctx.asScala.log).log(opts.level, "received message {}", msg)
+    target(ctx, msg)
+  }
```

Comment:
I noticed that `LoggingReceive` uses `isDefinedAt` to see if a message is handled an encodes that into the message.  I wasn't sure how to do that with the `Behavior` because I don't know how to access the underlying `PartialFunction`.

> We could actually even log what the new behavior/state is.

I'm not sure what you mean by this.

Reply:
It could be useful.  I can include it if you like.  

I have a question regarding the following line

> ```
> val handled = if (newBehavior == Behavior.UnhandledBehavior) "unhandled" else "handled"
> ```

This is checking if the `newBehavior` is a `Behavior.UnhandledBehavior`, but in `LoggingReceive` the logic is checking whether a message is actually handled by a behavior, which can be done using `PartialFunction.isDefinedAt`.  Is there a way to access the underlying PF of the behavior to do a similar check?

Reply:
In the typed case we don't know if it is a PF at all, so no, the only way to know that a message was unhandled is to call the behavior and look at the return value. Maybe we could skip that now and look into what would be useful in a follow up PR?

Reply:
Sounds good to me 👍

Reply:
Created #26271 to track that